pkg/reconciler/client.go

```diff
@@ -0,0 +1,124 @@
+package reconciler
+
+import (
+	clientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	customResourceDefinitionInformer "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions/apiextensions/v1"
+	internalinterfaces "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions/internalinterfaces"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/openservicemesh/osm/pkg/errcode"
+	"github.com/openservicemesh/osm/pkg/k8s"
+)
+
+// NewReconcilerClient implements a client to reconcile osm managed resources
+func NewReconcilerClient(kubeClient kubernetes.Interface, crdClient clientset.Interface, meshName string, stop chan struct{}, selectInformers ...k8s.InformerKey) error {
+	// Initialize client object
+	c := client{
+		kubeClient:      kubeClient,
+		meshName:        meshName,
+		apiServerClient: crdClient,
+		informers:       informerCollection{},
+	}
+
+	// Initialize informers
+	informerInitHandlerMap := map[k8s.InformerKey]func(){
+		CustomResourceDefinitions:    c.initCustomResourceDefinitionMonitor,
+		MutatingWebhookConfiguration: c.initMutatingWebhookConfigurationMonitor,
+	}
+
+	// If specific informers are not selected to be initialized, initialize all informers
+	if len(selectInformers) == 0 {
+		selectInformers = []k8s.InformerKey{CustomResourceDefinitions, MutatingWebhookConfiguration}
+	}
+
+	for _, informer := range selectInformers {
+		informerInitHandlerMap[informer]()
+	}
+
+	if err := c.run(stop); err != nil {
+		log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrStartingReconciler)).
+			Msg("Could not start osm reconciler client")
+		return err
+	}
+
+	return nil
+}
+
+// Initializes CustomResourceDefinition monitoring
+func (c *client) initCustomResourceDefinitionMonitor() {
+	osmCrdsLabel := map[string]string{constants.OSMAppNameLabelKey: constants.OSMAppNameLabelValue}
+
+	labelSelector := fields.SelectorFromSet(osmCrdsLabel).String()
+	options := internalinterfaces.TweakListOptionsFunc(func(opt *metav1.ListOptions) {
+		opt.LabelSelector = labelSelector
+	})
+
+	informerFactory := customResourceDefinitionInformer.NewFilteredCustomResourceDefinitionInformer(c.apiServerClient, k8s.DefaultKubeEventResyncInterval, cache.Indexers{nameIndex: metaNameIndexFunc}, options)
+
+	// Add informer
+	c.informers[CustomResourceDefinitions] = informerFactory
+
+	// Add event handler to informer
+	crdEventTypes := EventTypes{
+		Update: announcements.CRDUpdated,
+		Delete: announcements.CRDDeleted,
+	}
```

Comment:
You shouldn't need these event types with a custom event handler per resource, like in your implementation. Could you remove the event types initialized and also the announcements constants?